third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -0,0 +1,229 @@
+    if (newLayout != oldLayout) {
+      auto ty = cast<RankedTensorType>(expandType);
+      auto newTy = ty.cloneWithEncoding(newLayout);
+      builder.setInsertionPointAfter(expandDims);
+      src = builder.create<ttg::ConvertLayoutOp>(expandDims.getLoc(), newTy,
+                                                 expandDims);
+      copyAttrs(producer, src);
+    }
+    // Generate the store
+    Value trueVal = builder.create<arith::ConstantIntOp>(src->getLoc(), 1, 1);
+    auto storeOp = builder.create<ttng::TMEMStoreOp>(
+        src->getLoc(), allocOp, src->getResult(0), trueVal);
+    copyAttrs(producer, storeOp);
+  }
+
+  void TMEMLoad1D(Operation *producer, Operation *consumer) {
+    auto producerOutput = producer->getResult(0);
+    auto oldInputType = dyn_cast<RankedTensorType>(producerOutput.getType());
+    auto targetEncoding = oldInputType.getEncoding();
+    auto oldExpandType = getExpandedInput().getType();
+    Attribute newDistributedEncoding = ttng::getTmemCompatibleLayout(
+        oldExpandType.getShape()[0], oldExpandType.getShape()[1], oldExpandType,
+        numWarps);
+    auto newExpandType =
+        oldExpandType.cloneWithEncoding(newDistributedEncoding);
+    // Generate the load
+    builder.setInsertionPoint(consumer);
+    auto loadOp = builder.create<ttng::TMEMLoadOp>(
+        consumer->getLoc(), newExpandType,
+        builder.getType<ttg::AsyncTokenType>(), allocOp, Value());
+    copyAttrs(consumer, loadOp);
+    // Generate the reshape
+    auto reshape = builder.create<tt::ReshapeOp>(
+        consumer->getLoc(), oldInputType.getShape(), loadOp);
+    copyAttrs(consumer, reshape);
+    // Generate a convert layout.
+    auto newInput = builder.create<ttg::ConvertLayoutOp>(consumer->getLoc(),
+                                                         oldInputType, reshape);
+    copyAttrs(consumer, newInput);
+    // Replace the uses in the consumer
+    size_t numOperands = consumer->getNumOperands();
+    for (unsigned i = 0; i < numOperands; i++) {
+      if (consumer->getOperand(i) == producerOutput) {
+        consumer->setOperand(i, newInput);
+      }
+    }
+  }
+
+public:
+  void replaceWith1DTMEM(
+      Operation *producer, Operation *consumer,
+      std::optional<ttng::TMEMAllocOp> allocOpBuffer = std::nullopt) {
+    TMEMStore1D(producer, allocOpBuffer);
+    TMEMLoad1D(producer, consumer);
+  }
+};
+
+void generate1DAllocations(OpBuilder &builder, Block *globalStart,
+                           Operation *producer) {
+  assert(producer->hasAttr("tmem.start") && "Expected tmem.start");
+  auto producerPartition =
+      mlir::cast<mlir::IntegerAttr>(producer->getAttr("ttg.partition"))
+          .getInt();
+  for (auto consumer : producer->getUsers()) {
+    auto consumerParition =
+        mlir::cast<mlir::IntegerAttr>(consumer->getAttr("ttg.partition"))
+            .getInt();
+    if (producerPartition != consumerParition) {
+      TMEM1DAllocator(builder, globalStart)
```

Comment:
I realise this is just a test, but in practice would we create a new alloc and store for each consumer?